initial prometheus sketch

dpsutton opened the pull request:
TODO:

- [x] logging
- [x] better error message if env variable is unparseable as an integer
- [ ] ~ability for admin to turn off from admin page? (not enable, it requires an env variable, but might be nice to turn off if desired)~ cloud mentioned this is not necessary
- [x] research other metrics. Currently sending gc, memory, thread, and "standard" information.
- [x] handle failure to bind to prometheus port (addr already in use is easy way to see. also check privileged ports below 1000)
- [x] do we need instance-uuid to distinguish readings from different instances in a cluster? Answer: no. Measurements implicitly include and are grouped by the port they are scraped from. So two instances come from two different ports.

<img width="1441" alt="image" src="https://user-images.githubusercontent.com/6377293/187728487-cbbbe5dd-6cd4-4287-ad0f-2c385f3f467a.png">
<img width="1439" alt="image" src="https://user-images.githubusercontent.com/6377293/188999095-fb0aae03-754a-444f-9a08-08e97c8ec698.png">

#### Cluster groups by metric endpoint. Note here 9192 and 9193 ports for the two instances in the cluster:
<img width="1224" alt="image" src="https://user-images.githubusercontent.com/6377293/189209662-5253edc1-806e-4bc5-82af-bc0ecd716816.png">



#### Running prometheus
Download [here](https://prometheus.io/download)

make a config file that points to mb's metrics endpoint (for me is 9191)

```yaml
global:
  scrape_interval:     15s # By default, scrape targets every 15 seconds.

  # Attach these labels to any time series or alerts when communicating with
  # external systems (federation, remote storage, Alertmanager).
  external_labels:
    monitor: 'codelab-monitor'

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # Override the global default and scrape targets from this job every 5 seconds.
    scrape_interval: 5s

    static_configs:
      - targets: ['localhost:9191']
```

`./prometheus --config.file=prometheus.yml`

need env variable to start: MB_PROMETHEUS_SERVER_PORT=9191

sample of output:

```
'# HELP jvm_threads_current Current thread count of a JVM
'# TYPE jvm_threads_current gauge
jvm_threads_current 81.0
'# HELP jvm_threads_daemon Daemon thread count of a JVM
'# TYPE jvm_threads_daemon gauge
jvm_threads_daemon 36.0
'# HELP jvm_threads_peak Peak thread count of a JVM
'# TYPE jvm_threads_peak gauge
jvm_threads_peak 81.0
'# HELP jvm_threads_started_total Started thread count of a JVM
'# TYPE jvm_threads_started_total counter
jvm_threads_started_total 104.0
'# HELP jvm_threads_deadlocked Cycles of JVM-threads that are in deadlock waiting to acquire object monitors or ownable synchronizers
'# TYPE jvm_threads_deadlocked gauge
jvm_threads_deadlocked 0.0
```

request:

```
❯ http localhost:9191/metrics
HTTP/1.1 200 OK
Content-Length: 7329
Content-Type: text/plain; version=0.0.4; charset=utf-8
Date: Wed, 31 Aug 2022 16:13:38 GMT
Server: Jetty(9.4.48.v20220622)

'# HELP jvm_gc_collection_seconds Time spent in a given JVM garbage collector in seconds.
'# TYPE jvm_gc_collection_seconds summary
jvm_gc_collection_seconds_count{gc="G1 Young Generation",} 41.0
jvm_gc_collection_seconds_sum{gc="G1 Young Generation",} 0.586
jvm_gc_collection_seconds_count{gc="G1 Old Generation",} 0.0
jvm_gc_collection_seconds_sum{gc="G1 Old Generation",} 0.0
'# HELP jvm_threads_current Current thread count of a JVM
'# TYPE jvm_threads_current gauge
```

### Profiling

I'm running three jars:
- port 3006 is a jar from this branch with prometheus enabled
- port 3007 is a jar from this branch without prometheus enabled
- port 3008 is 0.44.2 release jar

For each running siege

```shell
  siege -c60 -r 20 "http://localhost:<port>/api/card/9/query POST" \
    -H 'Accept: application/json'  -H 'Content-Type: application/json' \
    -H 'Cookie: metabase.SESSION=<session-token>'
```

Prometheus enabled (port 3006):
Transaction rate:	      112.55 trans/sec

Prometheus present but not enabled (port 3007):
Transaction rate:	       92.95 trans/sec

0.44.2 release jar (no prometheus on classpath, StatisticsHandler on jetty):
Transaction rate:	       93.56 trans/sec

I don't think this sampling was scientific enough to think that things are faster, but we should be confident that we are in the same ballpark. 

Prometheus is sampling jvm stats and jetty stats. We do not seem to have introduced any performance issues with these metrics.

dpsutton committed:
initial prometheus sketch

need env variable to start: MB_PROMETHEUS_SERVER_PORT=9191

sample of output:

```
'# HELP jvm_threads_current Current thread count of a JVM
'# TYPE jvm_threads_current gauge
jvm_threads_current 81.0
'# HELP jvm_threads_daemon Daemon thread count of a JVM
'# TYPE jvm_threads_daemon gauge
jvm_threads_daemon 36.0
'# HELP jvm_threads_peak Peak thread count of a JVM
'# TYPE jvm_threads_peak gauge
jvm_threads_peak 81.0
'# HELP jvm_threads_started_total Started thread count of a JVM
'# TYPE jvm_threads_started_total counter
jvm_threads_started_total 104.0
'# HELP jvm_threads_deadlocked Cycles of JVM-threads that are in deadlock waiting to acquire object monitors or ownable synchronizers
'# TYPE jvm_threads_deadlocked gauge
jvm_threads_deadlocked 0.0
```

request:

```
❯ http localhost:9191/metrics
HTTP/1.1 200 OK
Content-Length: 7329
Content-Type: text/plain; version=0.0.4; charset=utf-8
Date: Wed, 31 Aug 2022 16:13:38 GMT
Server: Jetty(9.4.48.v20220622)

'# HELP jvm_gc_collection_seconds Time spent in a given JVM garbage collector in seconds.
'# TYPE jvm_gc_collection_seconds summary
jvm_gc_collection_seconds_count{gc="G1 Young Generation",} 41.0
jvm_gc_collection_seconds_sum{gc="G1 Young Generation",} 0.586
jvm_gc_collection_seconds_count{gc="G1 Old Generation",} 0.0
jvm_gc_collection_seconds_sum{gc="G1 Old Generation",} 0.0
'# HELP jvm_threads_current Current thread count of a JVM
'# TYPE jvm_threads_current gauge
```

dpsutton committed:
Log on unparseable prometheus port

dpsutton committed:
Clean up prometheus and save test

dpsutton committed:
typehint

dpsutton committed:
Jetty collector

qnkhuat commented on src/metabase/prometheus.clj:
```diff
@@ -0,0 +1,127 @@
+  [port registry]
+  (log/info (trs "Starting prometheus metrics web-server on port {0}" port))
+  (when-not port
+    (throw (ex-info (trs "Attempting to set up prometheus metrics web-server with no web-server port provided")
```
```suggestion
    (throw (ex-info (trs "Failed to set up prometheus metrics web-server with no web-server port provided")
```

IMO "Attempting" gives the impression of starting to setup, not something went wrong.

qnkhuat commented on src/metabase/prometheus.clj:
```diff
@@ -0,0 +1,127 @@
+  []
+  (when system
+    (try (stop-web-server system)
+         (log/info (trs "Prometheus web-server shut down"))
```
Should the' system be reset to `nil` after shutting down?

I can see this might make development that involves starting/stopping Prometheus more transparent. especially when the work involves configuring prometheus.

qnkhuat commented on src/metabase/server.clj:
```diff
@@ -95,9 +95,12 @@
-                    (* 10 60 1000))]
+                    (* 10 60 1000))
+        handler (async-proxy-handler handler timeout)
+        stats-handler (doto (StatisticsHandler.)
```
do we always want to initiate `statistichandler`?

I think we only want this if Prometheus is enabled right?

dpsutton replied:
We don't always want it. But I think its harmless enough that its better to not bring the notion of prometheus into the server/handler namespaces. If its not used it is harmless and there's way less branching in that namespace.

dpsutton committed:
Reset system to nil when shutting down

dpsutton committed:
c3p0 stats

dpsutton committed:
Clean up, document, and add tests

dpsutton committed:
Error message for failure to bind to port

Starting up with prometheus port set to the main webserver port to get
the error:

```
MB_JETTY_PORT=3006 MB_DB_CONNECTION_URI="postgres://..." \
MB_PROMETHEUS_SERVER_PORT=3006 java -jar locally-built.jar
```
yields the following error:

```shell
2022-09-09 10:08:52,000 INFO metabase.core :: Setting up prometheus metrics
2022-09-09 10:08:52,002 INFO metabase.prometheus :: Starting prometheus metrics collector
2022-09-09 10:08:52,016 INFO metabase.prometheus :: Starting prometheus metrics web-server on port 3,006
2022-09-09 10:08:52,036 ERROR metabase.core :: Metabase Initialization FAILED
clojure.lang.ExceptionInfo: Failed to initialized Prometheus on port 3,006 {:port 3006}
[stacktrace ...]
Caused by: java.io.IOException: Failed to bind to 0.0.0.0/0.0.0.0:3006
[stacktrace ...]
Caused by: java.net.BindException: Address already in use
```

dpsutton committed:
Test for error message

noahmoss commented on test/metabase/prometheus_test.clj:
```diff
@@ -0,0 +1,146 @@
+    "process_start_time_seconds"
+    "jetty_request_time_seconds_total"})
+
+(defn metric-tags
```
Should this and `metric-lines` be private?

dpsutton replied:
I've updated that. In general I don't care about test namespaces because there are no consumers that need to be directed to the api of a system or datastructure.

noahmoss reviewed (approved):
Implementation all looks reasonable to me; just a few minor comments

dpsutton committed:
Move prometheus to analytics folder

dpsutton committed:
Str port so log does not add commas to number

eg:
> Starting prometheus metrics web-server on port 9,191

dpsutton committed:
make some test functions private

dpsutton committed:
docstring on defsetting

dpsutton commented:
thanks for the review @noahmoss . I've moved the namespace to `metabase.analytics.prometheus`, cleaned up some docstrings and improved the log message. Thanks!

dpsutton commented on src/metabase/prometheus.clj:
```diff
@@ -0,0 +1,247 @@
+(ns metabase.prometheus
```
done

dpsutton commented on src/metabase/prometheus.clj:
```diff
@@ -0,0 +1,247 @@
+;; defsetting enables and [[system]] holds the system (webserver and registry)
+
+(defsetting prometheus-server-port
+  "Port to serve prometheus status from. If set"
```
fixed. thanks

qnkhuat commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+(ns metabase.analytics.prometheus
+  "Namespace for collection metrics with Prometheus. Will set up a registry and a webserver on startup
+  if [[prometheus-server-port]] is set to a port number. This can only be set in the environment and not though the
```
```suggestion
  if MB_PROMETHEUS_SERVER_PORT is set to a port number. This can only be set in the environment and not though the
```

Is it clearer if we specified the env var here instead of our setting?

dpsutton replied:
That's a good point. I'll add some more info there.

qnkhuat commented on test/metabase/analytics/prometheus_test.clj:
```diff
@@ -0,0 +1,146 @@
+        (is c3p0-collector "c3p0 stats not found"))))
+  (testing "Registry has an entry for each database in [[troubleshooting/connection-pool-info]]"
+    (with-prometheus-system [_ system]
+      (let [registry (.registry system)
```
these lets were not aligned.

qnkhuat commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+          (catch Exception e
+            (log/warn e (trs "Error stopping prometheus web-server")))))))
+
+(comment
```
Did you mean to keep this comment?

dpsutton replied:
I'll sometimes leave a comment that shows how to interact with the system a little bit. This one is instructive as a way to get a dump of the metrics without needing to hit the webserver.

qnkhuat reviewed (approved):
Overall this looks good.

I have some small styling comments.
Plus, do you think we need to document the env variable in our `docs/configuring-metabase/environment-variables.md`?

dpsutton committed:
align lets

dpsutton committed:
ns docstring changes

include the env variable for ease of understanding

dpsutton committed:
sort ns

dpsutton commented:
> Plus, do you think we need to document the env variable in our docs/configuring-metabase/environment-variables.md?

I think this is either automated or done by hand in the release preparation window. Either case we do not need to worry about it now. That is all collected prior to release. Good thought though!

dpsutton committed:
remove some reflection warnings

dpsutton committed:
reorder FE lines due to some new linter

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+
+(def ^:private label-translation
+  {:maxPoolSize        {:label       "c3p0_max_pool_size"
+                        :description "C3P0 Max pool size"}
```
Would it make sense to support translations here?

dpsutton replied:
oh yeah. maybe so.

dpsutton replied:
I'm doing it on the description. I think we want the keys to always remain the same?

metamben replied:
Yes, I think the keys should stay the same.

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+                      gauge (GaugeMetricFamily.
+                             ^String gauge-label
+                             ^String desc
+                             (->array ["database"]))]
```
```suggestion
                             (List/of "database"))]
```

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+   (completing conj!
+               (fn [metric-family-samples]
+                 (->array (persistent! metric-family-samples))))
+   stats))
```
I wonder if this function would be more readable as a loop adding elements to the mutable `ArrayList`. We are in the Java world anyway...

dpsutton replied:
```clojure
(defn stats->prometheus'
  [stats]
  (loop [[stat & remaining] stats
         arr                (ArrayList. (count stats))]
    (let [[raw-label measurements]               stat
          {gauge-label :label desc :description} (label-translation raw-label)]
      (if gauge-label
        (let [gauge (GaugeMetricFamily.
                     ^String gauge-label
                     ^String desc
                     (->array ["database"]))]
          (doseq [m measurements]
            (.addMetric gauge
                        (->array [(:label m)])
                        (:value m)))
          (.add arr gauge)
          (recur remaining arr))
        (recur remaining arr)))))
```
It's a bit shorter but I don't find it particularly easier to read. If you have a strong opinion I'll use it but I'm not seeing a great increase in readability.

metamben replied:
I had something like the following in mind (sorry, haven't tested):

```clojure
(defn stats->prometheus''
  [stats]
  (let [arr (ArrayList. (count stats))]
    (doseq [[raw-label measurements] stats
            :let [{gauge-label :label desc :description} (label-translation raw-label)]]
      (when gauge-label
        (let [gauge (GaugeMetricFamily.
                     ^String gauge-label
                     ^String desc
                     (List/of "database"))]
          (doseq [m measurements]
            (.addMetric gauge
                        (List/of (:label m))
                        (:value m)))
          (.add arr gauge))))
    arr))
```

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,248 @@
+  []
+  (when system
+    (locking #'system
+     (try (stop-web-server system)
```
Don't we want to handle several threads shutting down concurrently?

dpsutton replied:
ah right. I'll need another `when system` inside.

metamben reviewed (approved):
👍

dpsutton committed:
Cleanup: i18n descriptions, typo, List/of

dpsutton committed:
Merge branch 'master' into prometheus-initial-pass

dpsutton committed:
bit more concise c3p0 collection

dpsutton committed:
no longer need the helper `->array`

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -120,36 +120,22 @@
-   (comp (filter (fn [[raw-label _measurements]]
-                   (or (label-translation raw-label)
-                       (log/warn (trs "Unrecognized measurement {0} in prometheus stats"
-                                      raw-label)))))
```
This warning got lost in the translation. I'm not sure how important it is.

dpsutton replied:
good catch. It's not super important but would be nice to have the feedback that we don't track that measurement.

metamben commented on src/metabase/analytics/prometheus.clj:
```diff
@@ -0,0 +1,235 @@
+  [stats]
+  (let [arr (ArrayList. (count stats))]
+    (doseq [[raw-label measurements] stats
+            :let [{gauge-label :label desc :description} (label-translation raw-label)]
+            :when gauge-label
```
Compressing this in a `when-let` would be even nicer, IMHO.

dpsutton replied:
using an `if-let` made it simpler and gave a spot to bring the `log` back in. Good thinking.

dpsutton committed:
clean up doseq for c3p0 measurements
